endorsedLibraries/kotlinx.cli/src/main/kotlin/kotlinx/cli/Arguments.kt

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+package kotlinx.cli
+
+import kotlin.reflect.KProperty
+
+internal data class CLIEntityWrapper(var entity: CLIEntity<*, *>? = null)
+
+/**
+ * Command line entity.
+ *
+ * @property owner parser which owns current entity.
+ */
+abstract class CLIEntity<T : Any, TResult> internal constructor(val owner: CLIEntityWrapper) {
+    /**
+     * CLI element value.
+     */
+    internal lateinit var cliElement: ArgumentValueDelegate<TResult>
+    /**
+     * Wrapper  for element - read only property.
+     * Needed to close set of variable [cliElement].
+     */
+    val value: ArgumentValueDelegate<TResult>
+        get() = cliElement
+
+    /**
+     * Origin of argument value.
+     */
+    val valueOrigin: ArgParser.ValueOrigin
+        get() = (value as ParsingValue<*, *>).valueOrigin
+
+    abstract operator fun provideDelegate(thisRef: Any?, prop: KProperty<*>): ArgumentValueDelegate<TResult>
```

Comment:
The implementation looks the same in every inheritor. Can be lifted here.

Reply:
Fixed